forwardauth/forwardauth.go

```diff
@@ -0,0 +1,182 @@
+// Package forwardauth is a pomerium service that handles forward authentication requests.
+package forwardauth
+
+import (
+	"crypto/cipher"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+
+	envoy_service_auth_v2 "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	"github.com/gorilla/mux"
+
+	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/internal/encoding"
+	"github.com/pomerium/pomerium/internal/encoding/jws"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/sessions/cookie"
+	"github.com/pomerium/pomerium/internal/sessions/header"
+	"github.com/pomerium/pomerium/internal/sessions/queryparam"
+	"github.com/pomerium/pomerium/internal/urlutil"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+	"github.com/pomerium/pomerium/pkg/grpc"
+)
+
+const (
+	// ProxyTypeNginx is the name of nginx proxy.
+	ProxyTypeNginx = config.ProxyTypeNginx
+	// ProxyTypeTraefik is the name of traefik proxy.
+	ProxyTypeTraefik = config.ProxyTypeTraefik
+)
+
+// ProxyTypes contains all supported proxy types.
+var ProxyTypes = config.ProxyTypes
+
+const signinURL = "/.pomerium/sign_in"
+
+// ForwardAuth stores all the information associated with proxying a request.
+type ForwardAuth struct {
+	state          *atomicFaState
+	currentOptions *config.AtomicOptions
+	currentRouter  atomic.Value
+}
+
+// Mount mounts the authenticate routes to the given router.
+func (fa *ForwardAuth) Mount(r *mux.Router) {
+	r.PathPrefix("/").Handler(fa.registerFwdAuthHandlers())
```

Comment:
does this modify the existing router or create a new one?

Reply:
It modifies the input router, by creating new router (for current proxy type set in config) inside `registerFwdAuthHandlers`